<center>
<h4>Universidad Nacional de Córdoba - Facultad de Matemática, Astronomía, Física y Computación</h4>
<h3>Diplomatura en Ciencia de Datos, Aprendizaje Automático y sus Aplicaciones</h3>
</center>

<h1> Práctico I - Estadística </h1>
<h3> Análisis y Visualización de Datos - 2019 </h3>

Durante este práctico vamos a trabajar sobre el dataset [Human Freedom Index 2018](https://www.cato.org/human-freedom-index-new) de el instituto Cato. Este índice mide en detalle lo que entendemos como libertad, utilizando 79 indicadores de libertad personal y económica en distintos aspectos, hasta obtener un hermoso numerito del 1 al 10. Usaremos una [versión ya limpia del dataset](https://www.kaggle.com/gsutters/the-human-freedom-index/home) que pueden descargar desde Kaggle.

Las variables más importantes sobre las que trabaja el dataset son:

* Rule of Law
* Security and Safety
* Movement
* Religion
* Association, Assembly, and Civil Society
* Expression and Information
* Identity and Relationships
* Size of Government
* Legal System and Property Rights
* Access to Sound Money
* Freedom to Trade Internationally
* Regulation of Credit, Labor, and Business

Nosotros centrarermos nuestro análisis en variables relacionadas a *Identity and Relationships* en paises de Latinoamérica, y los compararemos con las estadísticas globales. La pregunta a responder es simple: **¿Qué niveles de libertad se viven en Latinoamérica, especificamente en cuanto libertades de indentidad?**. Sin embargo, para hacer un análisis de los datos tenemos que platear también estas sub preguntas:

1. ¿Qué significa tener un puntaje de 4.5? Hay que poner los puntajes de la región en contexto con los datos del resto del mundo.
2. ¿Cuál es la tendencia a lo largo de los años? ¿Estamos mejorando, empeorando?
3. En este estudio, la libertad se mide con dos estimadores principales: *hf_score* que hace referencia a Human Freedom, y *ef_score* que hace referencia a Economic Freedom. Estos dos estimadores, ¿se relacionan de la misma manera con la libertad de identidad?

Inicialmente, en toda exploración de datos tenemos muy poca información a priori sobre el significado de los datos y tenemos que empezar por comprenderlos. Les proponemos los siguientes ejercicios como guía para comenzar esta exploración.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn

In [2]:
seaborn.__version__

'0.9.0'

In [4]:
dataset = pd.read_csv('hfi_cc_2018.csv')
dataset.shape

(1458, 123)

In [5]:
dataset.columns  # Way too many columns!

Index(['year', 'ISO_code', 'countries', 'region', 'pf_rol_procedural',
       'pf_rol_civil', 'pf_rol_criminal', 'pf_rol', 'pf_ss_homicide',
       'pf_ss_disappearances_disap',
       ...
       'ef_regulation_business_bribes', 'ef_regulation_business_licensing',
       'ef_regulation_business_compliance', 'ef_regulation_business',
       'ef_regulation', 'ef_score', 'ef_rank', 'hf_score', 'hf_rank',
       'hf_quartile'],
      dtype='object', length=123)

Por suerte las columnas tienen un prefijo que nos ayuda a identificar a qué sección pertenecen. Nos quedamos sólo con las que comienzan con *pf_indentity*, junto con otras columnas más generales

In [6]:
important_cols = ['year', 'ISO_code', 'countries', 'region']

important_cols += [col for col in dataset.columns if 'pf_identity' in col]

important_cols += [
    'ef_score', # Economic Freedom (score)
    'ef_rank', # Economic Freedom (rank)
    'hf_score', # Human Freedom (score)
    'hf_rank', # Human Freedom (rank)
]

In [11]:
idataset = dataset[important_cols]
idataset

year ISO_code               countries                         region  \
0     2016      ALB                 Albania                 Eastern Europe   
1     2016      DZA                 Algeria     Middle East & North Africa   
2     2016      AGO                  Angola             Sub-Saharan Africa   
3     2016      ARG               Argentina  Latin America & the Caribbean   
4     2016      ARM                 Armenia        Caucasus & Central Asia   
5     2016      AUS               Australia                        Oceania   
6     2016      AUT                 Austria                 Western Europe   
7     2016      AZE              Azerbaijan        Caucasus & Central Asia   
8     2016      BHS                 Bahamas  Latin America & the Caribbean   
9     2016      BHR                 Bahrain     Middle East & North Africa   
10    2016      BGD              Bangladesh                     South Asia   
11    2016      BRD                Barbados  Latin America & the Caribbean   
12    2016      BLR                 Belarus                 Eastern Europe   
13    2016      BEL                 Belgium                 Western Europe   
14    2016      BLZ                  Belize  Latin America & the Caribbean   
15    2016      BEN                   Benin             Sub-Saharan Africa   
16    2016      BTN                  Bhutan                     South Asia   
17    2016      BOL                 Bolivia  Latin America & the Caribbean   
18    2016      BIH  Bosnia and Herzegovina                 Eastern Europe   
19    2016      BWA                Botswana             Sub-Saharan Africa   
20    2016      BRA                  Brazil  Latin America & the Caribbean   
21    2016      BRN       Brunei Darussalam                     South Asia   
22    2016      BGR                Bulgaria                 Eastern Europe   
23    2016      BFA            Burkina Faso             Sub-Saharan Africa   
24    2016      BDI                 Burundi             Sub-Saharan Africa   
25    2016      KHM                Cambodia                     South Asia   
26    2016      CMR                Cameroon             Sub-Saharan Africa   
27    2016      CAN                  Canada                  North America   
28    2016      CPV              Cape Verde             Sub-Saharan Africa   
29    2016      CAF       Central Afr. Rep.             Sub-Saharan Africa   
...    ...      ...                     ...                            ...   
1428  2008      SVN                Slovenia                 Eastern Europe   
1429  2008      ZAF            South Africa             Sub-Saharan Africa   
1430  2008      ESP                   Spain                 Western Europe   
1431  2008      LKA               Sri Lanka                     South Asia   
1432  2008      SDN                   Sudan             Sub-Saharan Africa   
1433  2008      SUR                Suriname  Latin America & the Caribbean   
1434  2008      SWZ               Swaziland             Sub-Saharan Africa   
1435  2008      SWE                  Sweden                 Western Europe   
1436  2008      CHE             Switzerland                 Western Europe   
1437  2008      SYR                   Syria     Middle East & North Africa   
1438  2008      TWN                  Taiwan                      East Asia   
1439  2008      TJK              Tajikistan        Caucasus & Central Asia   
1440  2008      TZA                Tanzania             Sub-Saharan Africa   
1441  2008      THA                Thailand                     South Asia   
1442  2008      TLS             Timor-Leste                     South Asia   
1443  2008      TGO                    Togo             Sub-Saharan Africa   
1444  2008      TTO     Trinidad and Tobago  Latin America & the Caribbean   
1445  2008      TUN                 Tunisia     Middle East & North Africa   
1446  2008      TUR                  Turkey     Middle East & North Africa   
1447  2008      UGA         

## 1. Estadísticos descriptivos

  1. Para comenzar con un pantallazo de los datos, calcular el rango de las variables.

In [12]:
# solo muestra variables cuantitativas
idataset.describe()

year  pf_identity_legal  pf_identity_parental_marriage  \
count  1458.000000         205.000000                     923.000000   
mean   2012.000000           6.307317                       7.248104   
std       2.582875           3.635095                       3.953387   
min    2008.000000           0.000000                       0.000000   
25%    2010.000000           7.000000                       5.000000   
50%    2012.000000           7.000000                      10.000000   
75%    2014.000000          10.000000                      10.000000   
max    2016.000000          10.000000                      10.000000   

       pf_identity_parental_divorce  pf_identity_parental  \
count                    923.000000           1358.000000   
mean                       7.632719              7.428203   
std                        3.689772              3.705913   
min                        0.000000              0.000000   
25%                        5.000000              5.000000   
50%                       10.000000             10.000000   
75%                       10.000000             10.000000   
max                       10.000000             10.000000   

       pf_identity_sex_male  pf_identity_sex_female  pf_identity_sex  \
count           1375.000000             1378.000000      1378.000000   
mean               6.709091                7.935414         7.325835   
std                4.683089                4.028877         4.048606   
min                0.000000                0.000000         0.000000   
25%                0.000000               10.000000         5.000000   
50%               10.000000               10.000000        10.000000   
75%               10.000000               10.000000        10.000000   
max               10.000000               10.000000        10.000000   

       pf_identity_divorce  pf_identity     ef_score      ef_rank  \
count           585.000000  1378.000000  1378.000000  1378.000000   
mean              7.538462     7.334180     6.785610    76.973149   
std               3.644142     3.159473     0.883601    44.540142   
min               0.000000     0.000000     2.880000     1.000000   
25%               5.000000     5.000000     6.250000    38.000000   
50%              10.000000     9.166667     6.900000    77.000000   
75%              10.000000    10.000000     7.410000   115.000000   
max              10.000000    10.000000     9.190000   162.000000   

          hf_score      hf_rank  
count  1378.000000  1378.000000  
mean      6.993444    77.007983  
std       1.025811    44.506549  
min       3.765827     1.000000  
25%       6.336685    39.000000  
50%       6.923840    76.000000  
75%       7.894660   115.000000  
max       9.126313   162.000000

In [14]:
# minimos y maximos de variables cuantitativas
idataset.describe().loc[['min','max']]

year  pf_identity_legal  pf_identity_parental_marriage  \
min  2008.0                0.0                            0.0   
max  2016.0               10.0                           10.0   

     pf_identity_parental_divorce  pf_identity_parental  pf_identity_sex_male  \
min                           0.0                   0.0                   0.0   
max                          10.0                  10.0                  10.0   

     pf_identity_sex_female  pf_identity_sex  pf_identity_divorce  \
min                     0.0              0.0                  0.0   
max                    10.0             10.0                 10.0   

     pf_identity  ef_score  ef_rank  hf_score  hf_rank  
min          0.0      2.88      1.0  3.765827      1.0  
max         10.0      9.19    162.0  9.126313    162.0

In [17]:
# obtengo el rango
v_range = idataset.describe().loc[['min','max']]
v_range.loc['range'] = v_range.loc['max'] - v_range.loc['min'] 
v_range.T # T: Transpose index and columns

min          max       range
year                           2008.000000  2016.000000    8.000000
pf_identity_legal                 0.000000    10.000000   10.000000
pf_identity_parental_marriage     0.000000    10.000000   10.000000
pf_identity_parental_divorce      0.000000    10.000000   10.000000
pf_identity_parental              0.000000    10.000000   10.000000
pf_identity_sex_male              0.000000    10.000000   10.000000
pf_identity_sex_female            0.000000    10.000000   10.000000
pf_identity_sex                   0.000000    10.000000   10.000000
pf_identity_divorce               0.000000    10.000000   10.000000
pf_identity                       0.000000    10.000000   10.000000
ef_score                          2.880000     9.190000    6.310000
ef_rank                           1.000000   162.000000  161.000000
hf_score                          3.765827     9.126313    5.360486
hf_rank                           1.000000   162.000000  161.000000

**Observación de los datos**

* El rango de años es de 8.
* Los rangos de las variables en las que nos enfoncamos (Personal Freedom) es de 10.
* Las variables ef_score y hf_score parecen corresponder a un promedio resultante de todas las variables de esos grupos, es decir ef y hf, su rango es de 6.31(ef) y 5.36(hf).
* Y por ultimo las variables ef_rank y hf_rank corresponde al ranking en que se ubica cada pais de los 162 paises en cuanto a su score de cada tipo(ef y hf) y su rango es 161.

2. Obtener media, mediana y desviación estándar de las variables *pf_identity* y *hf_score* en el mundo y compararla con la de Latinoamérica y el caribe. ¿Tiene sentido calcular la moda? 

  3. ¿Son todos los valores de *pf_identity*  y *hf_score* directamente comparables? ¿Qué otra variable podría influenciarlos?

  4. ¿Cómo pueden sanearse los valores faltantes?

  5. ¿Encuentra outliers en estas dos variables? ¿Qué método utiliza para detectarlos? ¿Los outliers, son globales o por grupo? ¿Los eliminaría del conjunto de datos?

## 2. Agregación de datos

1. Grafiquen la media de la variable *pf_identity* y *hf_score* a través de los años.
2. Realicen los mismos gráficos, pero separando por regiones (Cada variable en un gráfico distinto, sino no se ve nada). ¿La tendencia observada, es la misma que si no dividimos por regiones?
3. Si lo consideran necesario, grafiquen algunos países de Latinoamerica para tratar de explicar la tendencia de la variable *pf_identity* en la región. ¿Cómo seleccionarion los países relevantes a esa tendencia?

Hint: hay un gráfico de seaborn que hace todo por vos!

Sólo por curiosidad, graficar la tendencia de *hf_score* y *ef_score* a través de los años. ¿Tienen alguna hipótesis para este comportamiento?

## 2. Distribuciones
  1. Graficar en un mismo histograma la distribución de la variable *pf_identity* en global, y en Latinoamérica y el caribe. Repetir para la variable *hf_score*. ¿Visualmente, a qué tipo de distribución corresponde cada variable? ¿Es correcto utilizar todos los registros para esas zonas en estos gráficos?
  2. Realizar una prueba de Kolmogorov-Smirnof para comprobar analíticamente si estas variables responden la distribución propuesta en el ejercicio anterior. Hint: podés usar https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.kstest.html, pero hay que tener en cuenta que si la distribución es "norm", entonces va a comparar los datos con una distribución normal con media 0 y desviación estándar 1. Se puede utilizar la distribución sobre todos los datos o sólo sobre Latinoamérica.
  3. Realizar un gráfico QQ de las mismas distribuciones. Se puede utilizar a,bas distribuciones sobre todos los datos o sólo sobre Latinoamérica, pero no cruzadas.
  4. Medir la asimetría (skew) y curtosis de las mismas distribuciones con las que realizó el gráfico anterior. ¿Cómo se relacionan estos estadísticos con la forma del gráfico QQ obtenido previamente? ¿El gráfico QQ provee más información que no esté presente en estos estadísticos?

## 3. Correlaciones

En este ejercicio queremos responder a las preguntas

* Las libertades sociales y económicas, ¿van siempre de la mano?
* ¿Cómo se relacionan ambas con las libertades individuales y respectivas a las relaciones personales?

Para ello, analizaremos las correlaciones entre las variables pf_identity, hf_score y ef_score. 

Como pf_indentity contribuye al cálculo de hf_score y ef_score, esperamos hallar algún grado de correlación. Sin embargo, queremos medir qué tanta correlación. 

1. ¿Qué conclusiones puede sacar de un gráfico pairplot de estas tres variables? ¿Es adecuado para los valores de pf_identity? ¿Por qué?
2. Graficar la correlación entre pf_identity y hf_score; y entre pf_identity y ef_score. Analizar el resultado, ¿se pueden sacar conclusiones? Tengan en cuenta que como pf_identity es el resultado de un promedio, sólo toma algunos valores. Es, en efecto, discreta.
3. Calcular algún coeficiente de correlación adecuado entre los dos pares de variables, dependiendo de la cantidad de datos, el tipo de datos y la distribución de los mismo. Algunas opciones son: coeficiente de pearson, coeficiente de spearman, coeficientes de tau y de kendall. Interpretar los resultados y justificar si las variables están correlacionadas o no. 
4. [Opcional] Analizar la correlación entre la region y el hf_score (y/o el ef_score); y entre la region y el pf_identity. Considerar que como la variable *region* es ordinal, debe utilizarse algún tipo de test. Explicar cuáles son los requisitos necesarios para la aplicación de ese test. (Si no se cumplieran, se pueden agregar algunos datos para generar más registros)
